# ENCS5342 — Assignment #1
## Arabic Stemming, Lemmatization, and Regular Expressions

| | |
|---|---|
| **Course** | ENCS5342 — Information Retrieval with Applications of NLP |
| **Instructor** | Dr. Ahmed I. A. Shawahna — Birzeit University |
| **Semester** | Second Semester 2025–2026 |


| | Name | ID | Section |
|---|---|---|---|
| **Student 1** | leen Alqazaqi | 1220380 | Sec2 | 
| **Student 2** | Duha Imad | 1220623 | Sec2| 

---

## Grading & Submission

This assignment covers the use of **regular expressions** and **Python** for text processing,
and the application of **Arabic stemming and lemmatization** tools to a real multilingual corpus.

- Work in **groups of two students** (from any section).
- **Run all cells** in order before submitting. Output must be visible beneath each code cell.
- Where a question asks for discussion, add your written explanation in the **Markdown cell immediately below the code cell**.
- Submit a **single Jupyter Notebook file** named `A1_ID1_ID2.ipynb` (replace ID1 and ID2 with your student IDs).

---

## Before Starting

### Corpus: UN Parallel Corpus (UNCorpus)

This assignment uses the **United Nations General Assembly Resolutions Corpus (UNCorpus)**, a corpus of UN General Assembly resolutions. The UNCorpus is a six-language parallel text in Arabic, Chinese, English, French, Russian, and Spanish. The following paper describes the corpus:

> Rafalovitch, A. & Dale, R. (2009). *United Nations General Assembly Resolutions: A Six-Language Parallel Corpus.* Proceedings of MT Summit XII, Ottawa, Canada.  
> URL: http://web.science.mq.edu.au/~rdale/publications/papers/2009/MTS-2009-Rafalovitch.pdf

The portion of the UNCorpus used in this assignment is the file **`uncorpora_plain_20090831.tmx`**. This file is referred to as the **UNCorpus** throughout this document. It is stored in TMX format (Translation Memory eXchange), an XML-based standard for parallel corpora.

> **You can download the file from the following link:** https://drive.google.com/file/d/1AS1m3uzDL6B49xN4R7k2WWqOyRxP3QpI/view?usp=sharing  
> **Before running any cells:** place `uncorpora_plain_20090831.tmx` in the **same directory** as this notebook.

### Assignment Structure

| Section | Topic | Points |
|---------|-------|--------|
| **Section A** | Regular Expressions | 70 |
| **Section B** | Arabic Stemming and Lemmatization | 30 |
| | **Total** | **100** |


---
## Setup — Run This Cell First

This cell installs all required Python packages and loads the UNCorpus into memory. All subsequent cells depend on variables defined here. **Run this cell before any other.**

In [3]:
import re, os, sys, subprocess
from collections import Counter, defaultdict

# Install required packages
subprocess.run(
    [sys.executable, '-m', 'pip', 'install',
     'nltk', 'tashaphyne', 'sinatools', 'matplotlib',
     '-q', '--disable-pip-version-check'],
    capture_output=True
)

import nltk
import matplotlib.pyplot as plt
import matplotlib
matplotlib.rcParams['figure.dpi'] = 120
nltk.download('punkt', quiet=True)

# ── Load corpus ───────────────────────────────────────────────────
CORPUS_FILE = 'uncorpora_plain_20090831.tmx'

if not os.path.exists(CORPUS_FILE):
    print(f'ERROR: corpus file not found — {CORPUS_FILE}')
    print('Place the .tmx file in the same directory as this notebook.')
else:
    size_mb = os.path.getsize(CORPUS_FILE) / (1024 * 1024)
    with open(CORPUS_FILE, 'r', encoding='utf-8', errors='replace') as f:
        content   = f.read()
        f.seek(0)
        all_lines = f.readlines()
    print(f'Corpus file : {CORPUS_FILE}  ({size_mb:.1f} MB)')
    print(f'Lines loaded: {len(all_lines):,}')
    print('Setup complete.')


Corpus file : uncorpora_plain_20090831.tmx  (155.6 MB)
Lines loaded: 1,501,316
Setup complete.


---
# Section A: Regular Expressions (70 points)

All answers in this section must be obtained using **Python and the `re` module**. Use `re.search()`, `re.findall()`, or `re.sub()` as appropriate.

**How to present your answers:** Run the code cell for each question so that the output is visible beneath it, then write your written answer or discussion in the Markdown cell that follows.


---
### A0 (0 pt) — Worked Example: Lines Mentioning "Human Rights"

**Task:** Ignoring letter case, how many **lines** in the UNCorpus mention the phrase *Human Rights*?

This question is solved for you as a worked example illustrating the expected format: a code cell followed by a Markdown cell with the written answer.


In [4]:
# A0 — Worked example
count_hr = sum(1 for line in all_lines
               if re.search(r'human rights', line, re.IGNORECASE))
print(f'{count_hr:,} lines mention "Human Rights" (case-insensitive).')


5,664 lines mention "Human Rights" (case-insensitive).


**Answer:** 5,664 lines of text mention the phrase *Human Rights*.


---
### A1 (5 pt) — XML Structure of the UNCorpus

**Task:** Use Python and regular expressions to investigate and describe the XML structure of the UNCorpus TMX file. Your written answer must identify:

- The root element and all key container elements.
- Which element holds one complete translation unit (a sentence in all six languages).
- Which element holds one language's version of a sentence, and how the language is identified.
- Which element contains the actual text of a segment.
- Any metadata or property elements present in the file.

Include the regex patterns you used to discover each structural feature.


In [5]:
# A1 — write your code here — Investigate XML structure



**Answer:**

*(Describe the XML structure here based on the output above. Include the structure diagram and explain each element.)*

---
### A2 (10 pt) — Corpus-Level Statistics (2.5 pt each)

Answer each sub-question using Python and regular expressions on the full UNCorpus TMX file.
Run the code cell for each part and write the answer in the Markdown cell that follows.


#### A2a — How many lines does the UNCorpus file have?

In [6]:
# A2a — write your code here



**Answer:** *(write here)*


#### A2b — How many lines contain the segment tag `<seg>`?

In [7]:
# A2b — write your code here



**Answer:** *(write here)*


#### A2c — How many lines contain **no** `<seg>` tag?

These are lines with other tags such as `<prop>`, `</tuv>`, `<tu>`, `<header>`, etc.


In [8]:
# A2c — write your code here



**Answer:** *(write here)*


#### A2d — How many **English** segments does the UNCorpus contain?

In [9]:
# A2d — write your code here



**Answer:** *(write here)*


---
### A3 (25 pt) — The English UNCorpus

**Points:** 5 pt for the extraction setup | 4 pt each for sub-questions A3a–A3e

> **Definition:** A *word* in this section is any sequence of one or more **non-space** characters (i.e. tokens separated by whitespace). Punctuation and digits attached to a word are kept as part of it.

**Before answering A3a–A3e**, you must first extract the English plain text from the UNCorpus and save it to a file called `uncorpus.eng.txt`. This file must contain only the text content of the English segments — all XML tags must be removed.

Write your extraction code in the code cell below. Your code should follow these steps:

1. Iterate over the corpus lines and collect the `<seg>...</seg>` line that immediately follows each English TUV opening line.
2. Write the collected raw lines to `uncorpus.eng.txt`.
3. Read the file back, strip the `<seg>` and `</seg>` tags using `re.sub()`, and write the cleaned plain text back to the same file.
4. Verify your result using `os.path.getsize()`. The expected file size is **18,009,005 bytes**. If it does not match, revisit your line ending handling and tag stripping logic.

> *Note: On Linux/macOS you can also verify in a terminal with `wc -c uncorpus.eng.txt`, where `wc` stands for "word count" and the `-c` flag counts bytes.*

All sub-questions A3a–A3e use this file. At the bottom of your extraction cell, load the file into memory and split on whitespace to produce the token list used throughout A3.

In [10]:
# A3 Setup — EXACT implementation

eng_segments = []
#iteration over lines
for i in range(len(all_lines) - 1):
    if re.search(r'<tuv[^>]*xml:lang="EN"', all_lines[i], re.IGNORECASE):
        next_line = all_lines[i + 1]
        if re.search(r'<seg>.*</seg>', next_line):
            eng_segments.append(next_line.strip())
#write the collected raw lines
with open('uncorpus.eng.txt', 'w', encoding='utf-8') as f:
    for seg in eng_segments:
        f.write(seg + '\n')
#write the cleaned txt back
with open('uncorpus.eng.txt', 'r', encoding='utf-8') as f:
    text = f.read()
clean_text = re.sub(r'</?seg>', '', text)

with open('uncorpus.eng.txt', 'w', encoding='utf-8') as f:
    f.write(clean_text)
#verified length using `os.path.getsize()`
import os
print("File size:", os.path.getsize('uncorpus.eng.txt'))
#produce token list
with open('uncorpus.eng.txt', 'r', encoding='utf-8') as f:
    text = f.read()
token = text.split()

File size: 18009005


#### A3a — How many word tokens does `uncorpus.eng.txt` contain?

In [11]:
# A3a — write your code here
print(len(token))

2685538


**Answer:** *The file contains 2685538 word tokens*


#### A3b — How many unique word types does the file contain?

In [12]:
# A3b — write your code here
print(len(set(token)))


37032


**Answer:** *The file contains 37032 unique word types.*

#### A3c — How many unique word types remain after ignoring capitalization?

In [13]:
# A3c — write your code here
print(len(set(w.lower() for w in token)))

33364


**Answer:** *The file contains 33364 unique word types after ignoring capitalization by converting all words to lowercase.*

#### A3d — How many tokens consist **entirely of digits**? (e.g. `9000`)


In [26]:
# A3d — write your code here
digit_tokens = [w for w in token if re.findall(r'^\d+$', w)]
print(len(digit_tokens))

35857


**Answer:** *There are 35857 tokens that consist entirely of digits.*


#### A3e — How many tokens contain digits **and** at least one non-alphabetic character, but **no letters**? (e.g. `8,000.00` or `8-8`)

These are tokens that mix digits with punctuation characters such as commas, hyphens, or periods, but contain no letters.


In [27]:
# A3e — write your code here
tokens = [w for w in token if re.findall(r'\d', w) and not re.findall(r'[a-zA-Z]', w)]
print(len(tokens))

107194


**Answer:** *There are 107194 tokens that contain digits and non-alphabetic characters but no letters.*


---
### A4 (15 pt) — War and Peace in the UN (3 pt setup + 3 pt each)

This question uses `uncorpus.eng.txt`.
For all sub-questions, **ignore letter case and all non-alphabetic characters** — work with letters only.

**Before answering A4a–A4d**, write the setup code in the cell below. Your setup should:

1. Extract all alphabetic-only tokens from `uncorpus.eng.txt` using `re.findall()`.
2. Build a set of unique lowercased types from those tokens.
3. Print the total number of unique alphabetic word types as a sanity check. The expected number is 11,253 words.


In [16]:
# A4 Setup — write your code here



#### A4a — How many unique words **contain** the substring `war`?

In [17]:
# A4a — write your code here



**Answer:** *(write here)*


#### A4b — How many unique words **start with the substring** `war`?

In [18]:
# A4b — write your code here



**Answer:** *(write here)*


#### A4c — How many unique words **end with the substring** `war`?

In [19]:
# A4c — write your code here



**Answer:** *(write here)*


#### A4d — How many unique words **contain the substring** `war` but **neither start nor end** with it?


In [20]:
# A4d — write your code here


**Answer:** *(write here)*


---
### A5 (10 pt) — Word Frequency Histogram

This question uses `uncorpus.eng.txt`.

A **frequency histogram** is a list of all words sorted by how often they appear. It is one of the most fundamental tools in corpus linguistics and is the basis of Zipf's Law — the observation that word frequency drops sharply after the top few words, following a power-law distribution.

> **Definition:** A *word* here is any whitespace-separated token (keep all punctuation and digits; split on whitespace only).

**Tasks:**
1. Build the frequency table (word → count) for all tokens.
2. Report the **10 most frequent words** (top 10) with their counts and percentage of the corpus.
3. Report **10 hapax legomena** — words appearing exactly once. Report also the total hapax count and what fraction of the vocabulary they represent.
4. Plot a **bar chart** of the top 20 words by frequency.
5. **Discuss in 2–3 sentences:** What do the top 10 words have in common? What NLP concept does this illustrate?

**Points:** 3 pt for the top-10 frequency table | 3 pt for hapax analysis | 3 pt for the bar chart | 1 pt for discussion.


In [21]:
# A5 — write your code here — Word frequency histogram



In [22]:
# A5 — write your code here — Bar chart of top 20 words



**Discussion:** *(Write 2–3 sentences here.)*

---
### A6 (5 pt) — Top 20 Words per Language

Go back to the **original UNCorpus TMX file** for this question.

For each of the four languages — **English** (`en`), **Arabic** (`ar`), **Spanish** (`es`), and **Russian** (`ru`) — extract all segment text and find the **top 20 most frequent words**.

> **Definition:** A *word* is any whitespace-separated token. Keep all punctuation and digits; do **not** apply any regex filtering — split only on whitespace.

**Tasks:**
1. Display a top-20 frequency table for each language.
2. Plot a **bar chart** of the top 20 words for each language.
3. **Discuss in 2–3 sentences** per language: What do the top words reveal about that language's grammatical structure? (Consider function words, clitics, articles, and prepositions.)

**Points:** 2 pt for the four frequency tables | 2 pt for the four bar charts | 1 pt for discussion.


In [23]:
# A6 — write your code here — Top 20 words per language with bar charts



In [24]:
# A6 — write your code here — Bar charts for all four languages



**Discussion:** *(Write 2–3 sentences per language here.)*

---
# Section B: Arabic Stemming and Lemmatization (30 points)

This section works with the **Arabic segments** of the UNCorpus. You will extract the Arabic text using regular expressions (building on what you learned in Section A), initialize the stemming and lemmatization tools, and analyze the results.

**Tools you will use:**

| Tool | Type | How to import |
|------|------|---------------|
| NLTK ISRI | Root-based stemmer (no dictionary) | `from nltk.stem.isri import ISRIStemmer` |
| Tashaphyne | Light stemmer + root extractor | `from tashaphyne.stemming import ArabicLightStemmer` |
| Light10 | Light stemmer — affix stripping (Larkey et al., 2002) | Implemented below |
| Alma (SinaTools) | True lemmatizer + POS tagger — SinaLab, Birzeit University | `from sinatools.morphology import morph_analyzer` |


---
### B0 (0 pt) — Extract Arabic Text and Initialize Tools

Run this cell before attempting any question in Section B.

**Your task:** Write the regex pattern to extract all Arabic segment texts from the UNCorpus. Use the same approach as A3, but target the Arabic language code `"AR"` instead of `"EN"`. The rest of the cell — stemmer initialization and Alma setup — is provided for you.


In [25]:
# B0 — Extract Arabic segments and initialize all tools

# ── Your code: write the regex pattern to extract Arabic <seg> content ────────
# Hint: same structure as A3, but change "EN" to "AR"
ar_segments = re.findall(
    r'YOUR PATTERN HERE',  # <------ Modify me
    content, flags=re.DOTALL | re.IGNORECASE
)
ar_segments = [s.strip() for s in ar_segments if s.strip()]
ar_text     = ' '.join(ar_segments)

# Extract Arabic word tokens (Unicode Arabic block: U+0600–U+06FF)
ar_words = re.findall(r'[\u0600-\u06FF]+', ar_text)

print(f'Arabic segments extracted : {len(ar_segments):,}')
print(f'Arabic word tokens        : {len(ar_words):,}')
print(f'Arabic word types         : {len(set(ar_words)):,}')
print(f'Sample: {ar_segments[0][:100]}')

# ── Initialize stemmers (provided) ────────────────────────────────────────────
from nltk.stem.isri import ISRIStemmer
from tashaphyne.stemming import ArabicLightStemmer

isri_stemmer = ISRIStemmer()
tash_stemmer = ArabicLightStemmer()

class Light10Stemmer:
    """Reference implementation of the Light10 Arabic stemmer (Larkey et al., 2002)."""
    def __init__(self):
        self.prefixes   = ['وال', 'بال', 'كال', 'فال', 'لل', 'ال']
        self.suffixes   = ['ها', 'ان', 'ات', 'ون', 'ين', 'يه', 'ية', 'ه', 'ة', 'ي']
        self.diacritics = re.compile(r'[\u064B-\u0652\u0670]')
        self.hamza_map  = str.maketrans('أإآ', 'ااا')
    def stem(self, word):
        word = self.diacritics.sub('', word).translate(self.hamza_map)
        if len(word) > 3 and word[0] == 'و':
            word = word[1:]
        for p in sorted(self.prefixes, key=len, reverse=True):
            if word.startswith(p) and len(word) - len(p) >= 2:
                word = word[len(p):]; break
        for s in sorted(self.suffixes, key=len, reverse=True):
            if word.endswith(s) and len(word) - len(s) >= 2:
                word = word[:-len(s)]; break
        return word

light10 = Light10Stemmer()

# ── Alma (SinaTools) — downloads data files on first run ──────────────────────
ALMA_AVAILABLE = False
try:
    from sinatools.DataDownload.downloader import download_file, urls, get_appdatadir
    _app_dir  = get_appdatadir()
    _required = {
        'lemmas_dic.pickle':   'morph',
        'two_grams.pickle':    'two_grams',
        'three_grams.pickle':  'three_grams',
        'four_grams.pickle':   'four_grams',
        'five_grams.pickle':   'five_grams',
    }
    for fname, key in _required.items():
        if not os.path.exists(os.path.join(_app_dir, fname)):
            download_file(urls[key])
    from sinatools.morphology import morph_analyzer
    morph_analyzer.analyze('كتب')   # smoke test
    ALMA_AVAILABLE = True
    print('\nTools ready: ISRI | Tashaphyne | Light10 | Alma')
except Exception as e:
    print(f'\nTools ready: ISRI | Tashaphyne | Light10')
    print(f'Alma not available: {e}')
    print('Use the online demo: https://sina.birzeit.edu/alma/')

Arabic segments extracted : 0
Arabic word tokens        : 0
Arabic word types         : 0


IndexError: list index out of range

---
### B1 (8 pt) — Vocabulary Reduction by Stemming

A primary motivation for stemming in Information Retrieval is **vocabulary reduction**: grouping different surface forms of the same word under a single stem reduces the index size and improves recall by matching words that share a common root.

**Tasks:**
1. Apply **ISRI**, **Tashaphyne**, and **Light10** to all Arabic word tokens extracted in B0.
2. For each stemmer, compute the number of unique stems produced and the vocabulary reduction percentage relative to the original unstemmed vocabulary.
3. Display the results in a comparison table.
4. **Discuss in 3–4 sentences:** Which stemmer is most aggressive (reduces vocabulary the most)? Which is most conservative? What are the trade-offs between aggressiveness and precision for IR?


In [ ]:
# B1 — write your code here — Vocabulary reduction comparison



**Discussion:** *(Write 3–4 sentences here.)*

---
### B2 (7 pt) — Stemming Consistency: Do Stemmers Agree?

Different stemmers often produce different stems for the same word, which directly affects retrieval results — two documents about the same topic may not be matched if different stemmers normalize the query and document words differently.

**Tasks:**
1. Take the **50 most frequent Arabic word types** in the UNCorpus.
2. Apply **ISRI**, **Tashaphyne**, and **Light10** to each.
3. Display a table showing: word | frequency | ISRI stem | Tashaphyne stem | Light10 stem | agree?
4. Report how many of the top-50 words show full agreement across all three stemmers.
5. **Discuss in 2–3 sentences:** For words where the stemmers disagree, which stemmer produced the most useful stem for IR? Justify your answer with one specific example.


In [ ]:
# B2 — write your code here — Stemming consistency on top-50 Arabic words



**Discussion:** *(Write 2–3 sentences here.)*

---
### B3 (8 pt) — Stemming vs. Lemmatization on UN Political Terms

UN General Assembly resolutions use a specialized political and legal vocabulary. Stemmers may handle these technical terms differently from everyday words, and lemmatization — which maps each word to its dictionary headword — may be more appropriate for this domain.

**Tasks:** Apply all four tools to the following 15 politically significant Arabic terms that commonly appear in UN resolutions:

```
الأمم، المتحدة، الأمن، السلام، الحقوق، الإنسانية، القرار، التنمية،
الديمقراطية، الاحتلال، الإرهاب، المفاوضات، الاستقلال، السيادة، العدالة
```

1. Apply **ISRI**, **Tashaphyne**, and **Light10** to each word.
2. If Alma is available in your environment, apply it and include the diacritized lemma. Otherwise, look up each word at [sina.birzeit.edu/alma](https://sina.birzeit.edu/alma/) and fill in the Alma column manually.
3. Display a comparison table.
4. **Discuss in 3–4 sentences:** For a retrieval system operating on UN Arabic documents, which tool would you recommend and why? Address both precision (not merging unrelated words) and recall (grouping genuinely related word forms).


In [ ]:
# B3 — write your code here — Political terms: stemmers vs. lemmatizer
political_words = [
    'الأمم', 'المتحدة', 'الأمن', 'السلام', 'الحقوق',
    'الإنسانية', 'القرار', 'التنمية', 'الديمقراطية', 'الاحتلال',
    'الإرهاب', 'المفاوضات', 'الاستقلال', 'السيادة', 'العدالة',
]



**Discussion:** *(Write 3–4 sentences here.)*

---
### B4 (7 pt) — Over-Stemming in the Arabic UNCorpus

**Over-stemming** occurs when two or more words with different meanings are mapped to the same stem, causing the retrieval system to treat them as equivalent. This is a form of false positive: a document about one concept is retrieved in response to a query about a different concept, simply because they share a root.

**Tasks:**
1. Apply **ISRI stemming** to all Arabic word types in the UNCorpus.
2. Find all stems that are shared by **5 or more different original word types** — each such stem is a candidate for over-stemming.
3. Display the **top 10 most over-stemmed** stems: for each, show the stem, the number of original words that map to it, and up to 8 of those words.
4. **Discuss in 2–3 sentences:** Choose one stem from your top 10 and explain whether grouping all those words together would help or hurt a UN document retrieval system.


In [ ]:
# B4 — write your code here — Over-stemming analysis



**Discussion:** *(Write 2–3 sentences here.)*

---
## Grading Summary

### Section A — Regular Expressions (70 points)

| Question | Topic | Points | Your Points |
|----------|-------|--------|-------------|
| A0 | Worked example — Human Rights | 0 | |
| A1 | XML structure of the UNCorpus | 5 | |
| A2a | Number of lines in the corpus | 2.5 | |
| A2b | Lines containing `<seg>` | 2.5 | |
| A2c | Lines with no `<seg>` tag | 2.5 | |
| A2d | Number of English segments | 2.5 | |
| A3 setup | English plain-text extraction | 5 | |
| A3a | Total word tokens | 4 | |
| A3b | Unique word types (case-sensitive) | 4 | |
| A3c | Unique word types (case-insensitive) | 4 | |
| A3d | Digit-only tokens | 4 | |
| A3e | Mixed numeric tokens | 4 | |
| A4 setup | Alphabetic types extraction | 3 | |
| A4a–A4d | Substring matching for 'war' (3 pt each) | 12 | |
| A5 | Frequency histogram + bar chart + discussion | 10 | |
| A6 | Top-20 words per language + bar charts + discussion | 5 | |
| | **Section A Total** | **70** | |

### Section B — Arabic Stemming and Lemmatization (30 points)

| Question | Topic | Points | Your Points |
|----------|-------|--------|-------------|
| B0 | Extract Arabic text + initialize tools | 0 | |
| B1 | Vocabulary reduction by stemmer | 8 | |
| B2 | Stemming consistency — top-50 Arabic words | 7 | |
| B3 | Stemming vs. lemmatization on UN political terms | 8 | |
| B4 | Over-stemming analysis | 7 | |
| | **Section B Total** | **30** | |

### Final Grade

| Section | Points | Your Points |
|---------|--------|-------------|
| Section A — Regular Expressions | 70 | |
| Section B — Arabic Stemming and Lemmatization | 30 | |
| **Total** | **100** | |

---
**End of Assignment** | ENCS5342 — Information Retrieval with Applications of NLP | Birzeit University
